r2/r2/lib/media.py

```diff
@@ -205,6 +452,83 @@ def optimize_jpeg(filename):
     with open(os.path.devnull, 'w') as devnull:
         subprocess.check_call(("/usr/bin/jpegoptim", filename), stdout=devnull)
 
+class _SVGSoup(BeautifulSoup.BeautifulStoneSoup):
+    # Fixes unexpected conversion from `<g><g></g></g>` into `<g></g><g></g>`
+    nests = ('defs', 'g', 'marker', 'mask', 'pattern', 'svg', 'symbol', 'use')
+    NESTABLE_TAGS = BeautifulSoup.buildTagMap([], nests)
+
+def _sanitize_svg(image):
+    elements = [x.lower() for x in _SAFE_SVG_ELEMENTS]
+    attributes = [x.lower() for x in _SAFE_SVG_ATTRIBUTES]
+
+    # BS3 gets confused with some self-closing tags. We can safely 
+    # convert them with regex and afterward, we strip out unwanted <!> 
+    # and <?> tags.
+    #
+    # Keep in mind that `[^>]` goes in place of `.`, because it matches
+    # newline chars as well. One could also flag re.DOTALL, but we need
+    # to also not match `>`s.
+    image = re.sub(r'<([^>\s]+)\s*([^>]*)/>', r'<\1 \2></\1>',
+                   re.sub(r'<![^>]*>|<\?(?!xml)[^>]*>', r'', image))
+    soup = _SVGSoup(image, 
+                    convertEntities=BeautifulSoup.BeautifulStoneSoup.XML_ENTITIES)
+
+    inspect_element = soup.findAll(True)
+
+    for elem in inspect_element:
+        if elem.name not in elements: # Whitelist elements for security
+            if elem.contents and elem.parent:
+                parent = elem.parent
+                index = parent.contents.index(elem)
+
+                # Put all children in place of disallowed element
+                for child in elem.findAll(True, recursive=False):
+                    parent.insert(index + 1, child)
+                    inspect_element.append(child)
+                    index += 1
+            
+            elem.decompose()
+            continue
+
+        if elem.attrs:
+            to_del = set()
+            to_save = dict()
+
+            for attribute_set in elem.attrs:
+                attr = attribute_set[0]
+                value = attribute_set[1]
+                to_del.update({attr})
+
+                # When safe:
+                # Expression 1 == True
+                # Expression 2 == None
+                # Expression 3 == False
+
+                if (attr in attributes and
+                    re.search(r'url\((?!#)', elem[attr]) == None and
```

Comment:
There's a few ways around the `url(` check. Off the top of my head, Firefox lets you do things like `r"\u\r\l("`, and I seem to remember one browser allowing embedded nulls lik `b"ur\x00l("`.